File: python/78-Subsets.py
from typing import List
'''
Approach 1:
        0              N
     .     .        .     .
    1       N      1       N
   . .     . .    . .     . .
  2   N   2   N  2   N   2   N

Approach 2:
   0      1      2
 .   .    .  
1     2   2
.
2

Approach 3:
Generate binary representation of 0 ~ 7, then generate subsets from them
000, 001, 010, 011, 100, 101, 110, 111
'''

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        "Approach 1"
        n = len(nums)
        result = []
        def dfs(i, path):
            if i == n:
                result.append(path[:])
                return
            path.append(i)
            dfs(i+1, path)
            path.pop()
            dfs(i+1, path)
        dfs(0, [])
        return result
    def subsets(self, nums: List[int]) -> List[List[int]]:
        "Approach 2"
        n = len(nums)
        result = []
        def dfs(i, path):
            result.append(path[:])
            if i==n: return
            for j in range(i, n):
                path.append(nums[j])
                dfs(j+1, path)
                path.pop()
        dfs(0, [])
        return result
    # def subsets(self, nums: List[int]) -> List[List[int]]:
    #     "Approach 3"
    #     n = len(nums)
    #     ans = []
    #     for b in range(1<<n): # 0 ~ 2^n-1
    #         path = []
    #         for i in range(n):
    #             if b & 1<<i: # if ith bit is 1
    #                 path.append(nums[i])
    #         ans.append(path)
    #     return ans

nums = [1,2,3]
result = Solution().subsets(nums)
print(result) #[[], [1], [1, 2], [1, 2, 3], [1, 3], [2], [2, 3], [3]]]

nums = [0]
result = Solution().subsets(nums)
print(result) #[[], [0]]

File: cpp/78. Subsets.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
/* DFS 1:
        0              N
     .     .        .     .
    1       N      1       N
   . .     . .    . .     . .
  2   N   2   N  2   N   2   N
*/
public:
    vvi ans;
    void dfs(int i, vi &path, vi &nums) {
        if(i==nums.size()) {
            ans.push_back(path);
            return;
        }
        dfs(i+1, path, nums);
        path.push_back(nums[i]);
        dfs(i+1, path, nums);
        path.pop_back();
    }
    vvi subsets(vi& nums) {
        vi path;
        dfs(0, path, nums);
        return ans;
    }
};

class Solution {
/* DFS 2:
    0      1      2
  .   .    .  
 1     2   2
 .
 2
*/
public:
    vvi ans;
    void dfs(int i, vi &path, vi &nums) {
        int n = nums.size();
        ans.push_back(path);
        if(i==n) return;
        for(int j = i; j < n; j++) {
            path.push_back(nums[j]);
            dfs(j+1, path, nums);
            path.pop_back();
        }
    }
    vvi subsets(vi &nums) {
        vi path;
        dfs(0, path, nums);
        return ans;
    }
};

// class Solution {
// // Bit Manipulation
// public:
//     vvi subsets(vi& nums) {
//         int n = nums.size();
//         vvi ans;
//         for(int b = 0; b < 1<<n; b++) {
//             vi path;
//             for(int i = 0; i < n; i++) {
//                 if(b & 1<<i) path.push_back(nums[i]);
//             }
//             ans.push_back(path);
//         }
//         return ans;
//     }
// };

void printAns(vvi ans) {
    for(auto l: ans) {
        cout << "(";
        for(auto n: l) {
            cout << n << " ";
        }
        cout << "), ";
    }
    cout << "\n";
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; vvi ans;

    solution = Solution();
    nums = {1,2,3};
    ans = solution.subsets(nums);
    printAns(ans);

    solution = Solution();
    nums = {0};
    ans = solution.subsets(nums);
    printAns(ans);   
}